### PR TITLE
SQL DB Lock Exception added and IAM Cache Clean

### PR DESCRIPTION
**Pull request by mikechoch**
* IAM data will pile up on devices and down the road this could become an issue
  * Includes cleaning impressions, clicks, dismissals and SQL data for IAMs
* Db lock issue was present during opening db retries, related to #988
  * Added an extra exception to go along with SQLiteCantOpenDatabaseException, SQLiteDatabaseLockedException
* Replaced all `getWritableDatabase` and `getReadableDatabase` with `getSQLiteDatabase`
* Replaced all `getWritableDatabase` and `getReadableDatabase` with `getSQLiteDatabaseWithRetries`

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1007)

### CODE REVIEW

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java [hunk 1]
```diff
@@ -4,69 +4,188 @@
...
-                NotificationTable.COLUMN_NAME_CREATED_TIME + " < " + ((System.currentTimeMillis() / 1_000L) - 604_800L),
-                null);
+                cleanCachedNotifications(writableDb);
+                cleanCachedUniqueOutcomeEventNotifications(writableDb);
+            }
+
+            /**
+             * Deletes notifications with created timestamps older than 7 days
+             */
+            private void cleanCachedNotifications(SQLiteDatabase writableDb) {
```
Comment (Jeasmine) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java:53]: why are these methods (cleanCachedNotifications and cleanCachedUniqueOutcomeEventNotifications) inside the thread definition?

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java [hunk 2]
```diff
@@ -4,69 +4,188 @@
...
+                        OneSignalPrefs.PREFS_ONESIGNAL,
+                        OneSignalPrefs.PREFS_OS_IMPRESSIONED_IAMS,
+                        OSUtils.<String>newConcurrentSet());
+
+                Set<String> clickedClickIds = OneSignalPrefs.getStringSet(
+                        OneSignalPrefs.PREFS_ONESIGNAL,
+                        OneSignalPrefs.PREFS_OS_CLICKED_CLICK_IDS_IAMS,
+                        OSUtils.<String>newConcurrentSet());
+
+                if (oldMessageIds != null) {
```
Comment (Jeasmine) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java:160]: small optimization, but you can avoid dismissedMessages and impressionedMessages shared preference query if you already know that oldMessageIds is null or empty

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 3]
```diff
@@ -448,24 +435,21 @@ void onFailure(int statusCode, String response, Throwable throwable) {
      * For click counting, every message has it click id array
      * */
     private void setDataForRedisplay(OSInAppMessage message) {
-        if (!message.getDisplayStats().isRedisplayEnabled())
```
Comment (Jeasmine) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java:451]: why is this deleted?

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 4]
```diff
@@ -572,14 +556,10 @@ private void dismissCurrentMessage(@Nullable OSInAppMessage message) {
...
-    private void persistInAppMessageForRedisplay(final OSInAppMessage message) {
-        //If the IAM doesn't have the re display configuration then no need to save it
-        if (!message.getDisplayStats().isRedisplayEnabled())
-            return;
-
+    private void persistInAppMessage(final OSInAppMessage message) {
         long displayTimeSeconds = System.currentTimeMillis() / 1000;
-        message.getDisplayStats().setLastDisplayTime(displayTimeSeconds);
-        message.getDisplayStats().incrementDisplayQuantity();
+        message.getRedisplayStats().setLastDisplayTime(displayTimeSeconds);
```
Comment (Jeasmine) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java:561]: Sry I think you already explained me this, but why is always saving redisplay now?

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java [hunk 5]
```diff
@@ -444,6 +447,18 @@ private static void openURLInBrowser(@NonNull Uri uri) {
       return Collections.newSetFromMap(new ConcurrentHashMap<T, Boolean>());
    }
 
+   // Creates a new Set<String> from a Set String by converting and iterating a JSONArray
+   static Set<String> newStringSetFromString(String setString) throws JSONException {
```
Comment (Jeasmine) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java:451]: good one!

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java (hunk 4 above)
Comment (mikechoch) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java:561]: We should treat all IAMs the same way as it creates consistency 

Even though we arent visually seeing an IAM with a limit of 1, we still need to cache it and keep track of it. Its confusing saving only specific set of IAMs to SQL when you can easily do them all.

Specifically why I did this was because I needed to know when to clean Shared Preferences after 6 months and if we arent saving all IAMs we only end up the redisplay ones

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java (hunk 3 above)
Comment (mikechoch) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java:451]: This is why im always trying to rename things, we want to have redisplay data for IAMs so when we save them to SQL it matches the scheme and nothing is NULL

Redisplay isnt something we "enable" and "disable" also its just whether or not the IAM will show more than once. I think every IAM should have the same scheme and should be cached together, not separately

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRedisplayStats.java [hunk 6]
```diff
@@ -3,31 +3,31 @@
...
+    // Last IAM display time in seconds
     private long lastDisplayTime = -1;
-    //Current quantity of displays
+    // Current quantity of displays
     private int displayQuantity = 0;
-    //Quantity of displays limit
-    private int displayLimit = Integer.MAX_VALUE;
-    //Delay between displays in seconds
+    // Quantity of displays limit
+    private int displayLimit = 1;
```
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRedisplayStats.java:16]: Let's ensure the backend will always send this value for unlimited reshows.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java (hunk 5 above)
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java:451]: `setString` should be something like `jsonArray` instead.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java [hunk 7]
```diff
@@ -444,6 +447,18 @@ private static void openURLInBrowser(@NonNull Uri uri) {
...
+      Set<String> stringSet = new HashSet<>();
+
+      JSONArray clickIdsArray = new JSONArray(setString);
```
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java:454]: I think you forgot to rename `clickIdsArray` to something more generic when you moved it to OSUtils.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java [hunk 8]
```diff
@@ -4,69 +4,218 @@
...
+
+import org.json.JSONException;
+
+import java.util.Set;
 
 class OneSignalCacheCleaner {
 
-    private static String OS_DELETE_OLD_CACHED_DATA = "OS_DELETE_OLD_CACHED_DATA";
+    private final static long ONE_WEEK_IN_SECONDS = 604_800L;
+    private final static long SIX_MONTHS_IN_SECONDS = 15_552_000L;
```
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java:19]: `SIX_MONTHS_IN_SECONDS` should be something more descriptive to what is it used for like `IAM_MAX_CACHE_SEC`

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java [hunk 9]
```diff
@@ -4,69 +4,218 @@
...
+import android.support.annotation.WorkerThread;
+
+import org.json.JSONException;
+
+import java.util.Set;
 
 class OneSignalCacheCleaner {
 
-    private static String OS_DELETE_OLD_CACHED_DATA = "OS_DELETE_OLD_CACHED_DATA";
+    private final static long ONE_WEEK_IN_SECONDS = 604_800L;
```
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java:18]: `ONE_WEEK_IN_SECONDS` should be something more descriptive to what is it used. One week should be a comment instead.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java (hunk 8 above)
Comment (jkasten2) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java:19]: 6 months should be a comment instead.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRedisplayStats.java (hunk 6 above)
Comment (mikechoch) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRedisplayStats.java:16]: good UnitTest, on it

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java (hunk 5 above)
Comment (mikechoch) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java:451]: good idea, on it

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java (hunk 7 above)
Comment (mikechoch) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSUtils.java:454]: Whoops. thank you

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java (hunk 8 above)
Comment (mikechoch) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java:19]: You are correct, on it

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRedisplayStats.java (hunk 6 above)
Comment (Jeasmine) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRedisplayStats.java:16]: are you making this change on iOS too?

Comment (mikechoch) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRedisplayStats.java:16]: I will be but not yet

Comment (jkasten2) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRedisplayStats.java:16]: I did a quick check on the backed, looks like redisplay won't even get sent unless this value is set to 1 is what we want.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java [hunk 10]
```diff
@@ -4,69 +4,219 @@
...
      */
-    static void cleanOldUniqueOutcomeEventNotificationsCache(SQLiteDatabase writableDb) {
-        writableDb.delete(CachedUniqueOutcomeNotificationTable.TABLE_NAME,
-                "NOT EXISTS(SELECT NULL FROM " + NotificationTable.TABLE_NAME +
-                        " n WHERE" +
-                        " n." + NotificationTable.COLUMN_NAME_NOTIFICATION_ID  + " = " + CachedUniqueOutcomeNotificationTable.COLUMN_NAME_NOTIFICATION_ID + ")",
-                null);
+    private static void cleanCachedSharedPreferenceIamData(Set<String> oldMessageIds, Set<String> oldClickedClickIds) {
+        // IAMs without redisplay on with pile up and we need to clean these for dismissing, impressions, and clicks
+        if (oldMessageIds != null && oldMessageIds.size() > 0) {
```
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java:178]: This would be cleaner with 2 methods instead of long if blocks.

Comment (mikechoch) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java:178]: On it!
